posts/2017-05-02-io-monad-for-cats.md

```diff
@@ -0,0 +1,404 @@
+---
+layout: post
+title: An IO monad for cats
+
+meta:
+  nav: blog
+  author: djspiewak
+  pygments: true
+
+tut:
+  scala: 2.12.1
+  binaryScala: "2.12"
+  dependencies:
+    - org.scala-lang:scala-library:2.12.1
+    - org.typelevel::cats-effect:0.1-0848c9b
+---
+
+
+Haskell is a pure language.  Every Haskell expression is *referentially transparent*, meaning that you can substitute that expression with its evaluated result without changing the program.  Or, put into code:
+
+```haskell
+f e e
+-- is equivalent to
+let x = e in f x x
+```
+
```

Comment:
People get angry when we assume they understand Haskell. Can we make this a bit more clear, like

```haskell
-- this program
f expr expr -- apply function f to arguments expr, expr

-- is equivalent to this one, which factors out `expr`
let
  x = expr -- introduce a new binding `x`
in
  f x x
```